### ISSUE TITLE
large size of data using the model

### ISSUE THREAD

**Issue by FassyGit** [lyst/lightfm#217]
Hi, thank you for composing such a useful model to us.
I am recently working  on this model, and I tried several datasets on Foursquare. Some of them worked well, but when using the large dataset,  the model just kept running and gave no response. I have run it for almost five days on a cluster.
The information of the dataset I used is as follows:

> This dataset includes long-term (about 18 months from April 2012 to September 2013) global-scale check-in data collected from Foursquare. It contains 33,278,683 checkins by 266,909 users on 3,680,126 venues (in 415 cities in 77 countries). Those 415 cities are the most checked 415 cities by Foursquare users in the world, each of which contains at least 10K check-ins. Please see the references for more details about data collection and processing.

Can anyone help me on that?
What is the size of your runnable dataset

**Comment by maciejkula** [comment]
A dataset of this size should be easily runnable on a laptop in a couple of
hours.

Would you mind giving some details on the hyperparameters you used, and
also the details of your cluster? LightFM is single-machine only, so I
would be curious to see how you mesh it with distributed computing. In
particular, are you sure that you are spending time in LightFM rather than
shuffling data across the cluster?


On 15 Sep 2017 09:03, "FassyGit" <notifications@github.com> wrote:

Hi, thank you for composing such a useful model to us.
I am recently working on this model, and I tried several datasets on
Foursquare. Some of them worked well, but when using the large dataset, the
model just kept running and gave no response. I have run it for almost five
days on a cluster.
The information of the dataset I used is as follows:

This dataset includes long-term (about 18 months from April 2012 to
September 2013) global-scale check-in data collected from Foursquare. It
contains 33,278,683 checkins by 266,909 users on 3,680,126 venues (in 415
cities in 77 countries). Those 415 cities are the most checked 415 cities
by Foursquare users in the world, each of which contains at least 10K
check-ins. Please see the references for more details about data collection
and processing.

Can anyone help me on that?
What is the size of your runnable dataset

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/lyst/lightfm/issues/217>, or mute the thread

**Comment by FassyGit** [comment]
I used the parameters as mentioned in the document.

> NUM_THREADS = 4
> NUM_COMPONENTS = 100
> NUM_EPOCHS = 10
> ITEM_ALPHA = 1e-6
> 
> model = LightFM(loss='warp',
>                 item_alpha=ITEM_ALPHA,
>                 no_components=NUM_COMPONENTS)
> 
And as for the cluster problem, actually I am only using one node to run the program so there should not  be distribution problem here? Also runing other datasets on the cluster the same way as I do on thi s dataset works out the result fairly fast.
 And if you would like to read a little bit more. Here is the code:
`import codecs
import pandas as pd
import numpy as np
doc1 = codecs.open('dataset/dataset_TIST2015/dataset_TIST2015_Checkins.txt','rU','latin-1')
checkins = pd.read_csv(doc1, delimiter='\t')
checkins.columns=['userId', 'venueId', 'timeUTC', 'timeOffset']

doc2 = codecs.open('dataset/dataset_TIST2015/dataset_TIST2015_POIs.txt','rU','latin-1')
pois = pd.read_csv(doc2, delimiter='\t')
pois.columns=['venueId', 'lat', 'lon', 'venueCat', 'countryCode']


len(checkins['venueId'].unique())
len(checkins['userId'].unique())
len(pois['venueCat'].unique())

from sklearn.preprocessing import LabelEncoder
from scipy.sparse import csr_matrix

userIdencoder = LabelEncoder().fit(checkins['userId'])
checkins['userIdencoded'] = userIdencoder.transform(checkins['userId'])
n_users = len(userIdencoder.classes_)

venueIdencoder = LabelEncoder().fit(checkins.venueId)
checkins['venueIdEncoded'] = venueIdencoder.transform(checkins.venueId)
n_venues = len(venueIdencoder.classes_)

venueCatIdencoder = LabelEncoder().fit(pois.venueCat)
pois['venueCatIdEncoded'] = venueCatIdencoder.transform(pois.venueCat)
n_venueCats = len(venueCatIdencoder.classes_)


#
tags = venueCatIdencoder.classes_
tags.astype(dtype='<U50')

from sklearn.cross_validation import train_test_split
train_df, test_df = train_test_split(checkins, train_size = 0.8, random_state = 42)

train = csr_matrix((np.ones(train_df.shape[0]), (train_df.userIdencoded, train_df.venueIdEncoded)), shape=(n_users,n_venues))

test = csr_matrix((np.ones(test_df.shape[0]), (test_df.userIdencoded, test_df.venueIdEncoded)), shape=(n_users,n_venues))

from lightfm import LightFM
from lightfm.evaluation import auc_score
import time

NUM_THREADS = 4
NUM_COMPONENTS = 100
NUM_EPOCHS = 10
ITEM_ALPHA = 1e-6

model = LightFM(loss='warp',
                item_alpha=ITEM_ALPHA,
                no_components=NUM_COMPONENTS)

print ("I am beginning to model")
start = time.clock()

model.fit(train,epochs=NUM_EPOCHS,num_threads=NUM_THREADS)

print("model has been fitted")

elapse = (time.clock() - start)
print("Time used:", elapse)

train_auc = auc_score(model, train,num_threads=NUM_THREADS).mean()
print("Train_auc is %f" %train_auc)

test_auc = auc_score(model, test,train_interactions=train,num_threads=NUM_THREADS).mean()
print("Test_aus is %f" %test_auc)`

**Comment by maciejkula** [comment]
Thanks for the code (by the way, you can use triple backticks ``` to delimit code blocks).

1. So you are running it on a single machine. That's good, I presume there is nothing like Spark that would make things unnecessarily slow.
2. How many cores does your cluster machine have? 4 cores looks low; you should set `NUM_THREADS` to the number of available cores
3. In LightFM, evaluation is much more expensive than training (because in evaluation a score has to be computed for every item, for every user). I see you are computing the AUC score on the entire training set: I would expect that this takes the majority of the time. Did you ever get to the end of fitting? I would suggest you time just fitting first, and then experiment with smaller evaluation sets.

**Comment by FassyGit** [comment]
Thanks! It is exactly the problem you mentioned!
I did not take care of the core allocation problem, the default CPU number for every node is 2, so no wonder it is running slow.
After I fix this problem, it takes less than half an hour to fit the model.
And by the way, can you be more specific on the smaller evaluation sets? Does it mean to cut the train and test matrix after fitting the model?
Many thanks!

**Comment by maciejkula** [comment]
Great, I'm happy it helped.

As for the evaluation: when computing the accuracy metrics, LightFM will score all items for all users that are present (have at least one interaction) in your test set. So the time it takes to do an evaluation is proportional to the number of users in the test set. If it's taking too long, try limiting the number of users that have interactions in the test set.

Of course, this may reduce the value of your evaluation. However, beyond a certain number evaluating on more users has little benefit.
